src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java

```diff
@@ -15,135 +15,199 @@
-          ElasticsearchSinkConnectorConfig.BEHAVIOR_ON_NULL_VALUES_CONFIG
-      );
-    } catch (ConfigException e) {
-      throw new ConnectException(
-          "Couldn't start ElasticsearchSinkTask due to configuration error:",
-          e
-      );
+      // may be null if DLQ not enabled
+      reporter = context.errantRecordReporter();
+    } catch (NoClassDefFoundError | NoSuchMethodError e) {
+      // Will occur in Connect runtimes earlier than 2.6
+      log.warn("AK versions prior to 2.6 do not support the errant record reporter.");
     }
+
+    this.client = client != null ? client : new ElasticsearchClient(config, reporter);
+
+    log.info("Started ElasticsearchSinkTask.");
   }
 
   @Override
   public void put(Collection<SinkRecord> records) throws ConnectException {
-    log.debug("Putting {} records to Elasticsearch", records.size());
-    writer.write(records);
+    log.debug("Putting {} records to Elasticsearch.", records.size());
+    for (SinkRecord record : records) {
+      if (shouldSkipRecord(record)) {
+        logTrace("Ignoring {} with null value.", record);
+        reportBadRecord(record, new ConnectException("Cannot write null valued record."));
+        continue;
+      }
+
+      logTrace("Writing {} to Elasticsearch", record);
+
+      String index = convertTopicToIndexName(record.topic());
+      ensureIndexExists(index);
+      checkMapping(record);
+      tryWriteRecord(record);
+    }
   }
 
   @Override
   public void flush(Map<TopicPartition, OffsetAndMetadata> offsets) {
-    if (writer != null) {
-      log.debug("Flushing data to Elasticsearch with the following offsets: {}", offsets);
-      writer.flush();
-    } else {
-      log.debug("Could not flush data to Elasticsearch because ESWriter already closed.");
+    log.debug("Flushing data to Elasticsearch with the following offsets: {}", offsets);
+    try {
+      client.flush();
+    } catch (IllegalStateException e) {
+      log.debug("Tried to flush data to Elasticsearch, but BulkProcessor is already closed.");
     }
   }
 
   @Override
-  public void close(Collection<TopicPartition> partitions) {
-    log.debug("Closing the task for topic partitions: {}", partitions);
+  public void stop() {
+    log.debug("Stopping Elasticsearch client.");
+    client.close();
+    log.debug("Stopped Elasticsearch client.");
   }
 
   @Override
-  public void stop() throws ConnectException {
-    log.info("Stopping ElasticsearchSinkTask");
-    if (writer != null) {
-      writer.stop();
-      writer = null;
+  public String version() {
+    return Version.getVersion();
+  }
+
+  private void checkMapping(SinkRecord record) {
+    String index = convertTopicToIndexName(record.topic());
+    if (!config.shouldIgnoreSchema(record.topic()) && !existingMappings.contains(index)) {
+      if (!client.hasMapping(index)) {
+        client.createMapping(index, record.valueSchema());
+      }
+      log.debug("Locally caching mapping for index '{}'", index);
```

Comment:
nit: not a native speaker here, but maybe `Caching mapping for index '{}' locally`?